src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -599,6 +670,75 @@ private static ImmutableArray<RefKind> GetRefKindsOrNull(ArrayBuilder<RefKind> r
         }
 
+        // This fills in the arguments and parameters arrays in evaluation order.
+        private static void BuildExplicitArgumentsInEvaluationOrder(
+            SyntaxNode syntax,
+            Symbol methodOrIndexer,
+            bool expanded,
+            ImmutableArray<int> argsToParamsOpt,
+            ImmutableArray<BoundExpression> arguments,
+            Binder binder,
+            /* out */ ArrayBuilder<IArgument> argumentsBuilder,
+            /* out */ ArrayBuilder<ParameterSymbol> missingParametersBuilder)
+        {
+            ImmutableArray<ParameterSymbol> parameters = methodOrIndexer.GetParameters();
+            PooledHashSet<int> processedParameters = PooledHashSet<int>.GetInstance();
+
+            for (int a = 0; a < arguments.Length; ++a)
+            {
+                BoundExpression argument = arguments[a];
+
+                int p = (!argsToParamsOpt.IsDefault) ? argsToParamsOpt[a] : a;
+                var parameter = parameters[p];
+
+                Debug.Assert(!processedParameters.Contains(p));
+
+                processedParameters.Add(p); 
+
+                bool isParamArray = false;
+                if (expanded && p == parameters.Length - 1)
+                {
+                    isParamArray = true;
+
+                    ArrayBuilder<BoundExpression> paramArray = ArrayBuilder<BoundExpression>.GetInstance();
+                    paramArray.Add(argument);
+
+                    int remainingArgument = a + 1;
+                    for (; remainingArgument < arguments.Length; ++remainingArgument)
+                    {
+                        int remainingParameter = (!argsToParamsOpt.IsDefault) ? argsToParamsOpt[remainingArgument] : remainingArgument;
+                        if (remainingParameter != parameters.Length - 1)
+                        {
+                            break;
+                        }
+                        else
+                        {
+                            paramArray.Add(arguments[remainingArgument]);
+                        }
+                    }
+                    // Set loop variable to index of the last processed argument.
+                    a = remainingArgument - 1;
+
+                    var paramArrayType = parameters[p].Type;
+                    var arrayArgs = paramArray.ToImmutableAndFree();
+
+                    argument = CreateParamArrayArgument(syntax, paramArrayType, arrayArgs, null, binder);
+                }
+
+                argumentsBuilder.Add(BoundCall.CreateArgumentOperation(isParamArray ? ArgumentKind.ParamArray : ArgumentKind.Explicit, parameter, argument)); 
+            }
+
+            // Fill in parameters with missing arguments.
+            for (int i = 0; i < parameters.Length; ++i)
+            {
+                if (!processedParameters.Contains(i))
+                {
+                    missingParametersBuilder.Add(parameters[i]);
+                }
+            }
+            processedParameters.Free();
```

Comment:
Why return `missingParametersBuilder` rather than simply creating the missing parameters? In other words, would it make sense to inline `AppendMissingOptionalArguments` here?